File: terminal/src/com/jediterm/terminal/HyperlinkStyle.java
package com.jediterm.terminal;

import com.jediterm.terminal.model.hyperlinks.LinkInfo;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * @author traff
 */
public class HyperlinkStyle extends TextStyle {
  @NotNull
  private final LinkInfo myLinkInfo;

  @NotNull
  private final TextStyle myHighlightStyle;

  @Nullable
  private final TextStyle myPrevTextStyle;

  @NotNull
  private final HighlightMode myHighlightMode;

  public HyperlinkStyle(@NotNull TextStyle prevTextStyle, @NotNull LinkInfo hyperlinkInfo) {
    this(prevTextStyle.getForeground(), prevTextStyle.getBackground(), hyperlinkInfo, HighlightMode.HOVER, prevTextStyle);
  }

  public HyperlinkStyle(@Nullable TerminalColor foreground,
                        @Nullable TerminalColor background,
                        @NotNull LinkInfo hyperlinkInfo,
                        @NotNull HighlightMode mode,
                        @Nullable TextStyle prevTextStyle) {
    this(false, foreground, background, hyperlinkInfo, mode, prevTextStyle);
  }

  private HyperlinkStyle(boolean keepColors,
                         @Nullable TerminalColor foreground,
                         @Nullable TerminalColor background,
                         @NotNull LinkInfo hyperlinkInfo,
                         @NotNull HighlightMode mode,
                         @Nullable TextStyle prevTextStyle) {
    super(keepColors ? foreground : null, keepColors ? background : null);
    myHighlightStyle = new TextStyle.Builder()
        .setBackground(background)
        .setForeground(foreground)
        .setOption(Option.UNDERLINED, true)
        .build();
    myLinkInfo = hyperlinkInfo;
    myHighlightMode = mode;
    myPrevTextStyle = prevTextStyle;
  }

  @Nullable
  public TextStyle getPrevTextStyle() {
    return myPrevTextStyle;
  }

  @NotNull
  public TextStyle getHighlightStyle() {
    return myHighlightStyle;
  }

  @NotNull
  public LinkInfo getLinkInfo() {
    return myLinkInfo;
  }

  @NotNull
  public HighlightMode getHighlightMode() {
    return myHighlightMode;
  }

  @NotNull
  @Override
  public Builder toBuilder() {
    return new Builder(this);
  }

  public enum HighlightMode {
    ALWAYS, NEVER, HOVER
  }

  public static class Builder extends TextStyle.Builder {

    @NotNull
    private LinkInfo myLinkInfo;

    @NotNull
    private TextStyle myHighlightStyle;

    @Nullable
    private TextStyle myPrevTextStyle;

    @NotNull
    private HighlightMode myHighlightMode;

    private Builder(@NotNull HyperlinkStyle style) {
      myLinkInfo = style.myLinkInfo;
      myHighlightStyle = style.myHighlightStyle;
      myPrevTextStyle = style.myPrevTextStyle;
      myHighlightMode = style.myHighlightMode;
    }

    @NotNull
    public HyperlinkStyle build() {
      return build(false);
    }

    @NotNull
    public HyperlinkStyle build(boolean keepColors) {
      TerminalColor foreground = myHighlightStyle.getForeground();
      TerminalColor background = myHighlightStyle.getBackground();
      if (keepColors) {
        TextStyle style = super.build();
        foreground = style.getForeground() != null ? style.getForeground() : myHighlightStyle.getForeground();
        background = style.getBackground() != null ? style.getBackground() : myHighlightStyle.getBackground();
      }
      return new HyperlinkStyle(keepColors, foreground, background, myLinkInfo, myHighlightMode, myPrevTextStyle);
    }
  }
}

File: jediterm-core-mpp/src/jvmMain/kotlin/com/jediterm/terminal/HyperlinkStyle.kt
package com.jediterm.terminal

import com.jediterm.terminal.TextStyle.Builder
import com.jediterm.terminal.model.hyperlinks.LinkInfo

/**
 * @author traff
 */
class HyperlinkStyle private constructor(
    keepColors: Boolean,
    foreground: TerminalColor?,
    background: TerminalColor?,
    val linkInfo: LinkInfo,
    val highlightMode: HighlightMode,
    val prevTextStyle: TextStyle?
) : TextStyle(if (keepColors) foreground else null, if (keepColors) background else null) {
    val highlightStyle: TextStyle

    constructor(prevTextStyle: TextStyle, hyperlinkInfo: LinkInfo) : this(
        prevTextStyle.foreground,
        prevTextStyle.background,
        hyperlinkInfo,
        HighlightMode.HOVER,
        prevTextStyle
    )

    constructor(
        foreground: TerminalColor?,
        background: TerminalColor?,
        hyperlinkInfo: LinkInfo,
        mode: HighlightMode,
        prevTextStyle: TextStyle?
    ) : this(false, foreground, background, hyperlinkInfo, mode, prevTextStyle)

    init {
        this.highlightStyle = Builder()
            .setBackground(background)
            .setForeground(foreground)
            .setOption(Option.UNDERLINED, true)
            .build()
    }

    override fun toBuilder(): Builder {
        return Builder(this)
    }

    enum class HighlightMode {
        ALWAYS, NEVER, HOVER
    }

    class Builder internal constructor(style: HyperlinkStyle) : TextStyle.Builder() {
        private val myLinkInfo: LinkInfo

        private val myHighlightStyle: TextStyle

        private val myPrevTextStyle: TextStyle?

        private val myHighlightMode: HighlightMode

        init {
            myLinkInfo = style.linkInfo
            myHighlightStyle = style.highlightStyle
            myPrevTextStyle = style.prevTextStyle
            myHighlightMode = style.highlightMode
        }

        override fun build(): HyperlinkStyle {
            return build(false)
        }

        fun build(keepColors: Boolean): HyperlinkStyle {
            var foreground = myHighlightStyle.foreground
            var background = myHighlightStyle.background
            if (keepColors) {
                val style = super.build()
                foreground =
                    if (style.foreground != null) style.foreground else myHighlightStyle.foreground
                background =
                    if (style.background != null) style.background else myHighlightStyle.background
            }
            return HyperlinkStyle(keepColors, foreground, background, myLinkInfo, myHighlightMode, myPrevTextStyle)
        }
    }
}
